outlines/fsm/json_schema.py

```diff
@@ -191,362 +263,680 @@ def to_regex(
+        """
+        choices = [self.format_literal(choice) for choice in node["enum"]]
+        return self.format_anyOf(choices)
+
+    def visit_ref(self, node: Any, ctx: Context):
+        path = node["$ref"]
+        if path == "#":
+            raise NotImplementedError("Recursive schemas aren't supported")
+        new_node = self.resolver.lookup(path).contents
+        return self.to_regex(new_node, ctx)
+
+    def visit_allOf(self, node: Any, ctx: Context):
+        subpatterns = [self.to_regex(subschema, ctx) for subschema in node["allOf"]]
+        return self.format_allOf(subpatterns)
+
+    def visit_anyOf(self, node: Any, ctx: Context):
+        subpatterns = [self.to_regex(subschema, ctx) for subschema in node["anyOf"]]
+        return self.format_anyOf(subpatterns)
+
+    def visit_oneOf(self, node: Any, ctx: Context):
+        subpatterns = [self.to_regex(subschema, ctx) for subschema in node["oneOf"]]
+        return self.format_oneOf(subpatterns)
+
+    def visit_not(self, node: Any, ctx: Context):
+        raise NotImplementedError(
+            "`not` key in json schema isn't implemented. Please open an Outlines issue."
+        )
+
+    def visit_notimplemented(self, node: Any, ctx: Context):
+        raise NotImplementedError(
+            f"Handler for node `{node}` is not implemented. Please open an Outlines issue."
+        )
+
+    def format_number_range(
+        self,
+        min_digits_integer,
+        max_digits_integer,
+        min_digits_fraction,
+        max_digits_fraction,
+        min_digits_exponent,
+        max_digits_exponent,
+    ):
+        integers_quantifier = (
+            f"{{{min_digits_integer},{max_digits_integer}}}"
+            if min_digits_integer or max_digits_integer
+            else "*"
+        )
+        fraction_quantifier = (
+            f"{{{min_digits_fraction},{max_digits_fraction}}}"
+            if min_digits_fraction or max_digits_fraction
+            else "+"
+        )
+        exponent_quantifier = (
+            f"{{{min_digits_exponent},{max_digits_exponent}}}"
+            if min_digits_exponent or max_digits_exponent
+            else "+"
+        )
+        return rf"((-)?(0|[1-9][0-9]{integers_quantifier}))(\.[0-9]{fraction_quantifier})?([eE][+-][0-9]{exponent_quantifier})?"
+
+    def format_integer_range(self, min_digits=None, max_digits=None):
+        if min_digits or max_digits:
+            num_items_pattern = f"{{{min_digits},{max_digits}}}"
+        else:
+            num_items_pattern = "*"
+
+        return rf"(-)?(0|[1-9][0-9]{num_items_pattern})"
+
+    def format_string_length(self, min_length, max_length):
+        return f'"{STRING_INNER}{{{min_length},{max_length}}}"'
+
+    def format_string_pattern(self, pattern: str):
+        return f'"{pattern[1:-1] if pattern[0] == "^" and pattern[-1] == "$" else pattern}"'
+
+    def format_string_format(self, fmt: str):
+        format_regex = format_to_regex.get(fmt)
```

Comment:
We should remove this from the module scope and make it exclusively a class member.